program.c

```diff
@@ -172,3 +172,26 @@ int program_find_bootable_partition(void)
 
 	return part;
 }
+
+void program_unload(void)
+{
+	struct program *program;
+
+	programes_last = NULL;
+
+	for (program = programes; program; program = program->next) {
+		if (program->filename)
+			xmlFree(program->filename);
```

Comment:
Analog to standard free() it's valid to pass NULL to this function, so remove the extra conditionals before each one of these.